pkg/apis/batch/validation/validation.go

```diff
@@ -120,6 +120,7 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
 func ValidateJob(job *batch.Job, opts JobValidationOptions) field.ErrorList {
 	// Jobs and rcs have the same name validation
 	allErrs := apivalidation.ValidateObjectMeta(&job.ObjectMeta, true, apivalidation.ValidateReplicationControllerName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateJobName(job.ObjectMeta.Name)...)
```

Comment:
Two things:

1) the way to do this is to pass the validator to `ValidateObjectMeta` above - it currently passes `apivalidation.ValidateReplicationControllerName` which is an alias to `apimachineryvalidation.NameIsDNSSubdomain`.  You probably want `apimachineryvalidation.NameIsDNSLabel`.

2) IIUC, this is a breaking change - Jobs which used to pass validation will no longer pass, and if any such Job is saved in the API, it might not be updateable.  The only way this is acceptable is if you can show that *all* such Jobs could not possibly have worked, under any circumstances.

Is that true?  I thought the "manualSelector" feature meant that a user COULD create a long-name Job and use their own selector and it would be OK?

So at best, you have to check the "manualSelector" field here, I think.

Here's how I would sequence the issues I flagged:

1) Fix the registry logic so that validate does not mutate the object.  Move the selector generation to PrepareForCreate() and fix up the tests.  This sets the stage for...

2) Add test cases for long names, both with and without manual selectors.  I expect this will fail after the above, but with an unintuitive error.  Fix validation so the name validation is correct.

3) Fix up the default value setting for `job.Spec.ManualSelector == nil` -> `false`.  Find the places that check for nil and simplify.

Reply:
Sounds like https://github.com/kubernetes/kubernetes/issues/116137 should be resolved first before this PR.